consensus/hotstuff/eventhandler/event_handler_v2.go

```diff
@@ -0,0 +1,395 @@
+
+	// this is a sanity check to see if the block is really for the current view.
+	curView := e.paceMaker.CurView()
+	if block.View != curView {
+		return fmt.Errorf("sanity check fails: block proposal's view does not match with curView, (blockView: %v, curView: %v)",
+			block.View, curView)
+	}
+
+	// checking if I'm the next leader
+	nextView := curView + 1
+	nextLeader, err := e.committee.LeaderForView(nextView)
+	if err != nil {
+		return fmt.Errorf("failed to determine primary for next view %d: %w", nextView, err)
+	}
+
+	return e.processBlockForCurrentViewWithNextLeader(block, e.committee.Self(), nextLeader)
+}
+
+func (e *EventHandlerV2) processBlockForCurrentViewWithNextLeader(block *model.Block, self flow.Identifier, nextLeader flow.Identifier) error {
+
+	log := e.log.With().
+		Uint64("block_view", block.View).
+		Hex("block_id", block.BlockID[:]).
+		Uint64("parent_view", block.QC.View).
+		Hex("parent_id", block.QC.BlockID[:]).
+		Hex("signer", block.ProposerID[:]).
+		Logger()
+
+	// voter performs all the checks to decide whether to vote for this block or not.
+	isSelfNextLeader := self == nextLeader
+	curView := block.View
+
+	shouldVote := true
+	ownVote, err := e.voter.ProduceVoteIfVotable(block, curView)
+	if err != nil {
+		if !model.IsNoVoteError(err) {
+			// unknown error, exit the event loop
+			return fmt.Errorf("could not produce vote: %w", err)
+		}
+		log.Debug().Err(err).Msg("should not vote for this block")
+		shouldVote = false
+	}
+
+	// send my vote if I should vote if I'm not the leader for the current view
+	if shouldVote {
+		e.notifier.OnVoting(ownVote)
+		log.Debug().Msg("forwarding vote to compliance engine")
+
+		// if I'm the next leader, just forward the vote to the vote aggregator of ourselves
+		// if I'm not the next leader, then send the vote to the next leader.
+		if isSelfNextLeader {
+			// TODO: consider add own vote along with the proposal atomically to VoteAggregator
+			// in order to ensure the own vote will always be included.
+			err := e.voteAggregator.AddVote(ownVote)
+			log.Warn().Err(err).Msg("could not forward vote to the vote aggregator of ourselves")
+		} else {
+			err := e.communicator.SendVote(ownVote.BlockID, ownVote.View, ownVote.SigData, nextLeader)
+			if err != nil {
+				log.Warn().Err(err).Msg("could not forward vote")
+			}
+		}
+	}
+
+	// inform pacemaker that we've done the work for the current view, it should increment the current view
+	_, viewChanged := e.paceMaker.UpdateCurViewWithBlock(block, isSelfNextLeader)
+	if !viewChanged {
+		// this is a sanity check
+		// when I've processed the block for the current view and I'm not the next leader,
+		// the pacemaker should trigger a view change
+		nextView := curView + 1
+		return fmt.Errorf("pacemaker should trigger a view change to (view %v), but didn't", nextView)
+	}
+
+	// current view has changed, go to new view
+	return e.startNewView()
```

Comment:
## ⚠️ missing logic for primary to collect votes (before view change)

we should only process to next view, if the PaceMaker tells us to